### PR TITLE
Added ability to filter latest posts by author

### PR DESCRIPTION
**Pull request by pstonier**
## Description
Added dropdown to select author to filter the latest posts in the Latest Posts block.

## How has this been tested?
Locally.

## Screenshots 
Screen Shot 2019-06-13 at 4 50 27 PM
Screen Shot 2019-06-13 at 4 50 39 PM

## Types of changes
- Added AuthorSelect Component to QueryControls based on the CategorySelect Component
- Added `users` attribute to LatestPosts
- LatestPosts now fetches a user list to populate the dropdown like categories 

## Checklist:
- [ X ] My code is tested.
- [ X ] My code follows the WordPress code style. 
- [ X ] My code follows the accessibility standards. 
- [ X ] My code has proper inline documentation. 
- [ n/a ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by talldan** [comment]
Thanks for working on this @pstonier!

The PR was discussed in a core-editor triage (https://wordpress.slack.com/archives/C02QB2JS7/p1562825386362400 - link requires slack sign up)

We decided to switch the label to Needs Design Feedback. Firstly we felt technical feedback can be provided through the normal code review process. Secondly, it'd be good to understand the motivation behind this PR (as there's no linked issue) and to make sure the UI is implemented in the right way in terms of design and behaviour of the dropdown.

**Comment by pstonier** [comment]
All sounds good. From the motivation factor: we had been working on a page that was a bio about some of our consultants and needed a way to list the articles they'd authored. I had expected it to be an available filter option and when I found that it wasn't, I found my way around to adding it.

[Example Page](https://salienthealthcare.com/about/ryan-mackman/)

**Review by draganescu — COMMENTED** [review]
This PR effectively copies the same logic as the category filter and from my tests it appears that it works correctly. I do not approve only because of the "Needs design feedback" label, but my opinion is that we go ahead with it.

**Comment by paaljoachim** [comment]
This feature seems straight forward to me. 

Can some design folks add a comment please? As it would be nice to get it code reviewed and merged. @karmatosed @mapk @kjellr @jasmussen @melchoyce

**Comment by jasmussen** [comment]
About to do some travelling so can't review in depth. But based on the screenshots alone, this seems good. It's an advanced feature, so it's in the sidebar, and it's a dropdown like the others. 👍 👍

**Comment by paaljoachim** [comment]
This PR looks ready for a final code review and merge. Thanks!
@talldan

File: packages/block-library/src/latest-posts/edit.js [hunk 1]
```diff
@@ -34,13 +34,17 @@ import { withSelect } from '@wordpress/data';
 const CATEGORIES_LIST_QUERY = {
 	per_page: -1,
 };
+const USERS_LIST_QUERY = {
+	per_page: -1,
```
Comment (aprea) [inline packages/block-library/src/latest-posts/edit.js]: Wonder if this it going to scale okay? Just trying to think what'll happen on a site with hundreds of thousands or millions of users.

Comment (aprea) [reply packages/block-library/src/latest-posts/edit.js]: Oh, never mind me just read the REST pagination docs.
> Large queries can hurt site performance, so per_page is capped at 100 records. If you wish to retrieve more than 100 records, for example to build a client-side list of all available categories, you may make multiple API requests and combine the results within your application.
https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/

**Comment by draganescu** [comment]
In the mean time I've added #20415 implementing the same thing, with the hope of getting a faster course :) @pstonier would you look and either move this or that one towards a merge?

**Comment by mapk** [comment]
@draganescu what's the verdict here? Which PR are we going with?

**Comment by draganescu** [comment]
Will review this one today

File: packages/block-library/src/latest-posts/block.json [hunk 2]
```diff
@@ -12,6 +12,12 @@
 		"categories": {
 			"type": "array"
 		},
+		"users": {
+			"type": "array"
+		},
+		"items": {
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/block.json:18]: What do we need this items for? It seems that you needed it for the array type, but it should be nested into users. Now that I're recommented switching to number it is not needed at all.

File: packages/block-library/src/latest-posts/block.json [hunk 3]
```diff
@@ -12,6 +12,12 @@
 		"categories": {
 			"type": "array"
 		},
+		"users": {
+			"type": "array"
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/block.json:16]: This makes the PR not work in my tests. Given that we only filter by one author and all we need is the ID which is a Number, we should have this set to "number"

File: packages/block-library/src/latest-posts/edit.js [hunk 4]
```diff
@@ -262,6 +280,13 @@ class LatestPostsEdit extends Component {
 						categorySuggestions={ categorySuggestions }
 						onCategoryChange={ selectCategories }
 						selectedCategories={ categories }
+						onAuthorChange={ ( value ) =>
+							setAttributes( {
+								users: '' !== value ? value : undefined,
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:285]: We should make sure we save the ID as number, it is string by default, so:
`users: '' !== value ? Number( value ) : undefined`

**Review by draganescu — COMMENTED** [review]
Howdy @pstonier this didn't work out of the box. I left some comments about what I had to change. Other than that, great work, we're almost done!

File: packages/block-library/src/latest-posts/block.json (hunk 2 above)
Comment (pstonier) [reply packages/block-library/src/latest-posts/block.json]: Items was an attempt to resolve the issue of the error: "Notice: Undefined index: items in /app/public/wp-includes/rest-api.php on line 1245, but as you mentioned. It's not necessary and I've removed it.

I've also switched the type to number on users and is working as expected.

File: packages/block-library/src/latest-posts/block.json (hunk 3 above)
Comment (pstonier) [reply packages/block-library/src/latest-posts/block.json]: I agree. I've updated the type to number.

File: packages/block-library/src/latest-posts/edit.js (hunk 4 above)
Comment (pstonier) [reply packages/block-library/src/latest-posts/edit.js]: I agree. I've updated the code.

File: packages/block-library/src/latest-posts/edit.js [hunk 5]
```diff
@@ -97,6 +114,7 @@ class LatestPostsEdit extends Component {
 			order,
 			orderBy,
 			categories,
+			users,
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:117]: As above, let's update this to author.

File: packages/block-library/src/latest-posts/edit.js [hunk 6]
```diff
@@ -262,6 +280,14 @@ class LatestPostsEdit extends Component {
 						categorySuggestions={ categorySuggestions }
 						onCategoryChange={ selectCategories }
 						selectedCategories={ categories }
+						onAuthorChange={ ( value ) =>
+							setAttributes( {
+								users:
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:285]: This as well.

File: packages/block-library/src/latest-posts/edit.js [hunk 7]
```diff
@@ -459,6 +485,7 @@ export default withSelect( ( select, props ) => {
 		order,
 		orderBy,
 		categories,
+		users,
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:488]: Same, update to author.

File: packages/block-library/src/latest-posts/edit.js [hunk 8]
```diff
@@ -470,6 +497,7 @@ export default withSelect( ( select, props ) => {
 	const latestPostsQuery = pickBy(
 		{
 			categories: catIds,
+			author: users,
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:500]: we can change to shorthand here:
 ```
{
	categories: catIds,
	author: users,
...
```

File: packages/block-library/src/latest-posts/index.php [hunk 9]
```diff
@@ -49,6 +49,9 @@ function render_block_core_latest_posts( $attributes ) {
 	if ( isset( $attributes['categories'] ) ) {
 		$args['category__in'] = array_column( $attributes['categories'], 'id' );
 	}
+	if ( isset( $attributes['users'] ) ) {
+		$args['author'] = $attributes['users'];
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/index.php:53]: `$args['author'] = $attributes['author'];`

**Review by draganescu — CHANGES_REQUESTED** [review]
Hey @pstonier we're 99% there, only one request left, which I should have spotted the last time I reviewed, that is using author instead of users as an attribute (and hence updating all over accordingly). It makes more sense, both the singular and the name change.

File: packages/block-library/src/latest-posts/block.json [hunk 10]
```diff
@@ -12,6 +12,9 @@
 		"categories": {
 			"type": "array"
 		},
+		"users": {
```
[the remark this answers is not in the document]
Comment (pstonier) [reply packages/block-library/src/latest-posts/block.json]: I'm good with this.

**Review by mapk — APPROVED** [review]
I just tested this today and it's working as expected. I like that I'm able to filter the posts by author now! Thanks. I'm approving from a design perspective.

One thing to note is now that I can filter by Author, I'd like a toggle to show the author which should probably exist here:

Screen Shot 2020-05-05 at 8 05 15 AM

Of course, don't let that hold up this PR.

**Comment by pstonier** [comment]
> I just tested this today and it's working as expected. I like that I'm able to filter the posts by author now! Thanks. I'm approving from a design perspective.
> 
> One thing to note is now that I can filter by Author, I'd like a toggle to show the author which should probably exist here:
> 
> Screen Shot 2020-05-05 at 8 05 15 AM
> 
> Of course, don't let that hold up this PR.

I like that idea and have updated the code to add that.

File: packages/block-library/src/latest-posts/edit.js [hunk 11]
```diff
@@ -411,6 +446,12 @@ class LatestPostsEdit extends Component {
 										__( '(no title)' )
 									) }
 								</a>
+								{ displayAuthor && (
+									<div className="wp-block-latest-posts__post-author">
+										{ __( 'Posted by' ) }{ ' ' }
```
Comment (draganescu) [inline packages/block-library/src/latest-posts/edit.js:451]: 👋 This should be incorporating the author's name like in #20595 

```JS
<span className="wp-block-latest-posts__post-author">
	{ sprintf(
		/* translators: byline. %s: current author.
		__( 'by %s' ),
		currentAuthor.name
	) }
</span>
```

**Comment by draganescu** [comment]
@pstonier @mapk  #20595 adds the "Show post author" toggle in the inspector.

File: packages/block-library/src/latest-posts/edit.js (hunk 11 above)
Comment (pstonier) [reply packages/block-library/src/latest-posts/edit.js]: I agree this is a better approach. I've updated my code accordingly.

**Comment by mapk** [comment]
When the code is dialed in again and you'd like another review, just ping me. @draganescu thanks for stepping in to help.

**Comment by draganescu** [comment]
Howdy @pstonier ,

Trying to test this, setting show author option I get:

```JS
TypeError: post.author_info is undefined edit.js:453:15
```

and the block crashes. 

Also a rebase would be in order.

**Comment by pstonier** [comment]
> Howdy @pstonier ,
> 
> Trying to test this, setting show author option I get:
> 
> ```js
> TypeError: post.author_info is undefined edit.js:453:15
> ```
> 
> and the block crashes.
> 
> Also a rebase would be in order.

Thanks. I have fixed this and rebased.

**Review by draganescu — APPROVED** [review]
I have tested this and works pretty well. Since we're only adding new data to the block we shouldn't need a deprecation function, testing a previously created block with this PR updated with no issues.

Great job @pstonier !

**Comment by pstonier** [comment]
Thank you for all the help!

**Pull request merged**